clang/include/clang/Basic/AttrDocs.td

```diff
@@ -455,6 +455,174 @@ The SYCL kernel in the previous code sample meets these expectations.
   }];
 }
 
+def SYCLKernelEntryPointDocs : Documentation {
+  let Category = DocCatFunction;
+  let Content = [{
+The ``sycl_kernel_entry_point`` attribute facilitates the generation of an
+offload kernel entry point, sometimes called a SYCL kernel caller function,
+suitable for invoking a SYCL kernel on an offload device. The attribute is
+intended for use in the implementation of SYCL kernel invocation functions
+like the ``single_task`` and ``parallel_for`` member functions of the
+``sycl::handler`` class specified in section 4.9.4, "Command group ``handler``
+class", of the SYCL 2020 specification.
+
+The attribute requires a single type argument that specifies a class type that
+meets the requirements for a SYCL kernel name as described in section 5.2,
+"Naming of kernels", of the SYCL 2020 specification. A unique kernel name type
+is required for each function declared with the attribute. The attribute may
+not first appear on a declaration that follows a definition of the function.
+
+The attribute only appertains to functions and only those that meet the
+following requirements.
+
+* Has a ``void`` return type.
+* Is not a non-static member function, constructor, or destructor.
+* Is not a C variadic function.
+* Is not a coroutine.
+* Is not defined as deleted or as defaulted.
+* Is not declared with the ``constexpr`` or ``consteval`` specifiers.
```

Comment:
`constexpr` would only be meaningful here if the SYCL invocation functions like `single_task` were also declared `constexpr`. I think this would be worth reconsidering when/if an effort appears to enable `constexpr` support in the SYCL library.

Reply:
I don't see a good reason to limit it here, I think we should just implement it as support.  Particularly as the library isn't necessarily the only user of this.  Plus, it allows users to USE existing library functions.  That is, it is an artificial restriction on users who wnat to re-use existing code, even if we don't ever execute at constexpr time.

Reply:
I think there is some misunderstanding here. Can you provide an example that illustrates what you have in mind?

These functions aren't useful on their own; a call to them is specified as a no-op and is done for the side effect of triggering emission of a kernel entry point function (any ODR-use will do for that; it doesn't have to be a call). I think it would be quite strange if a call in a constexpr context evaluated the body of the function but a call outside did not. From the documentation in the PR:
> The call to ``kernel_entry_point()`` has no effect other than to trigger emission of the entry point function. The statments that make up the body of the function are not executed when the function is called; they are only used in the generation of the entry point function.

Reply:
The body of the `kernel_entry_point` is still required to be valid C++, and often just call the function, unless they are replaced at code-gen time (at which point, what you quote is correct).  

We'd received quite a few bugs/reports/wishes that these functions could be called in a constexpr context such that their body would just be executed in that context (since the body actually DOES the correct evaluation).  One such use case was folks trying to not use the SYCL library, but instead just augment their existing program to get 'free' parallelism, (plus some alterations because of our limitations on the shape of the function) and then being disappointed because they actually have valid constexpr uses.

If the SYCL library doesn't choose to do this (and the onus is on the author of the function to ensure that the 'as-if' rule gets satisfied by the body) they don't have to, but the limit here is artificial IMO.

Reply:
Just rethinking about this, since I have worked today on `constexpr` in a related context https://github.com/KhronosGroup/SYCL-CTS/pull/976
This looks like a pessimistic interpretation of the specification or something we could clarify in the SYCL committee.
It is not really important to consider `consteval` as it will always be in a constant evaluated context, so it will never reach SYCL codegen (perhaps in the future if we want to speed up the slooow constant evaluator of Clang itself on some accelerators? :rofl: ).
And why not having a kernel which is `constexpr` function too? It will only matter when in a non-constant evaluated context.

Reply:
>And why not having a kernel which is constexpr function too? It will only matter when in a non-constant evaluated context.

That is sorta my thoughts too.  `consteval` doesn't really make sense:  We COULD do it, but it would be a no-op attribute, which I think is reason enough to diagnose it.  

`constexpr` on the other hand definitely makes sense to me.

Reply:
With regard to `constexpr`, consider the following example. Is the lambda actually invoked when `f()` is called?
```
template<typename KN, typename KT>
[[clang::sycl_kernel_entry_point(KN)]]
void skep(KT k) { k(); }
void f() {
  skep<struct K>([]{});
}
```

The answer is no. As described in the documentation and quoted in an earlier comment in this conversation, the call to the `sycl_kernel_entry_point` attributed function happens, but it's apparent body is not actually executed. The body serves as a pattern for the offload kernel entry point function.

With that in mind, what should the following do?
```
constexpr auto v = (skep<struct K>([]{}), 0);
```

The answer is that it should do nothing (invoking the lambda would be inconsistent with what happens outside of constant evaluation).

The way to support constant evaluation of a SYCL kernel (which may be `constexpr`) would be to write the SYCL invocation function to detect constant evaluation and react accordingly.
```
template<typename KN, typename KT>
constexpr void single_task(KT k) {
  if consteval {
    k();
  } else {
    skep<KN>(k); // Trigger emission of the offload kernel entry point.
    // FIXME: Do stuff to invoke the offload kernel entry point
  }
}
constexpr auto v = (single_task<struct K>([]{}), 0); // Huzzah, constant evaluated SYCL kernel.
```

Reply:
That is what I was talking about that the author of said function ends up being responsible for the as-if rule.

Reply:
Thanks, Erich. I'd still like to understand what your mental model is for what `constexpr` would mean on one of these functions; just to ensure there isn't some misunderstanding.

I've been assuming that the model you had in mind would effectively implicitly inline the `if consteval` statement from my `single_task()` example into `skep()` such that it would behave when called as-if written as follows:
```
template<typename KN, typename KT>
[[clang::sycl_kernel_entry_point(KN)]]
void skep(KT k) {
  if consteval {
    k();
  } else {
    // Do nothing; an offload entry point function will have been generated
    // that the SYCL run-time library will arrange to call.
  }
}
```

This is the reason for my comments about inconsistent behavior in and out of constant evaluation context. Note that any caller would still have to be written to be sensitive to whether invocation is occurring during constant evaluation since it would have to arrange (or not) for the offload entry point to be invoked on the appropriate device.

If you have a different model in mind, please do share details.

Reply:
That is effectively  my mental model, except without the `if consteval`, since the body of the `skep` doesn't really exist when doing the offload codegen. The point is there is a lot of `device` stuff that gets arranged that is, in my mental model, implementation details for "call this function object, just faster/more parallel". 

However, the 'faster/more parallel' is a bit of a QoI thing, so 'call this slower, but at compile time' should work as well.

Reply:
Ok, thanks. Note that this design doesn't require the `#ifdef __SYCL_DEVICE_ONLY__` dance that was needed with the `sycl_kernel` attribute. This is an intentional divergence made to accommodate a future single pass compilation model. So, the body of `skep()` does exist during both host and device compilation, it just isn't evaluated when called. These are details for a later PR.

> However, the 'faster/more parallel' is a bit of a QoI thing, so 'call this slower, but at compile time' should work as well.

I agree, but that needs to be handled at a higher level. `sycl_kernel_entry_point` attributed functions can't fill the role of a user facing interface since they don't include the mechanism needed for device dispatch. These functions can only be used as part of the implementation detail for higher level interfaces.

Reply:
After some background thoughts, this `constexpr` aspect is only for the runtime function itself and more generally the SYCL runtime is currently far from being `constexpr`. So, we can just prevent `constexpr` and `consteval` and change this implementation detail in the future if this becomes a problem.
Note that this does not prevent to have the kernel lambda to be `constexpr` today.

Reply:
Thank you, @keryell, yes, that matches my perspective. I'll go ahead and resolve this conversation now. If you have no further concerns, could you please approve this PR?